### PR TITLE
Add weighted fair resource group scheduling policy

### PR DESCRIPTION
**Pull request by raghavsethi**
This policy allocates query slots amongst eligible resource groups in
realtime, weighted by shares. Of all competing groups, the group with
the lowest ratio of actual utilization (running queries/total queries)
to expected utilization (share/sum of shares of all eligible groups)
of the cluster will win the slot. If multiple groups have the same
ratio, a lottery is held where the number of tickets for each group is
proportional to the share.

Will add documentation if the approach/naming looks good.

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroup.java [hunk 1]
```diff
@@ -430,13 +432,13 @@ public int getSchedulingWeight()
     }
 
     @Override
-    public void setSchedulingWeight(int weight)
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroup.java]: Unnecessary change? `weight` seemed like a perfectly valid name for the parameter.

File: presto-spi/src/main/java/com/facebook/presto/spi/resourceGroups/SchedulingPolicy.java [hunk 2]
```diff
@@ -17,5 +17,6 @@
 {
     FAIR,
     WEIGHTED,
+    WEIGHTED_SHARES,
```
Comment (martint) [inline presto-spi/src/main/java/com/facebook/presto/spi/resourceGroups/SchedulingPolicy.java]: We should look into whether we can remove the other policies.

Comment (raghavsethi) [reply presto-spi/src/main/java/com/facebook/presto/spi/resourceGroups/SchedulingPolicy.java]: Not sure about that, we made resource groups as non experimental in the last release. IMO these do need to be renamed - this is actually a fair scheduler with weights. We would then have: `FCFS, STOCHASTIC, FAIR_SHARE, QUERY_PRIORITY`.

File: presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroup.java (hunk 1 above)
Comment (raghavsethi) [reply presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroup.java]: Unintended, yes.

File: presto-spi/src/main/java/com/facebook/presto/spi/resourceGroups/SchedulingPolicy.java (hunk 2 above)
Comment (martint) [reply presto-spi/src/main/java/com/facebook/presto/spi/resourceGroups/SchedulingPolicy.java]: Well, I don't think we should be accumulating algorithms either. As we find ones that work better, we should be deprecating and removing the old ones.

File: presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/FairSharingQueue.java [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+    @Override
+    public Iterator<E> iterator()
+    {
+        return index.keySet().iterator();
+    }
+
+    public static class Usage
+    {
+        private int share;
+        private int utilization;
```
Comment (nezihyigitbasi) [inline presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/FairSharingQueue.java]: please add a comment that explains what these fields mean.

File: presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/FairSharingQueue.java [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+
+        public Usage(int share, int utilization)
+        {
+            checkArgument(share > 0, "share must be positive");
+            checkArgument(utilization >= 0, "usage must be zero or positive");
```
Comment (nezihyigitbasi) [inline presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/FairSharingQueue.java]: `usage` -> `utilization`

File: presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/FairSharingQueue.java [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+        }
+
+        public E getValue()
+        {
+            return value;
+        }
+
+        public void update(Usage usage)
+        {
+            this.usage = usage;
```
Comment (nezihyigitbasi) [inline presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/FairSharingQueue.java]: `requireNonNull`

File: presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/FairSharingQueue.java [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+            if (candidateFactor <= winnerFactor) {
+                if (candidateFactor < winnerFactor) {
+                    winnerFactor = candidateFactor;
+                    winners.clear();
+                }
+                winners.add(candidate);
+            }
+        }
+
+        return pickWinner(winners);
```
Comment (nezihyigitbasi) [inline presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/FairSharingQueue.java]: this can return `null`, is this OK? Please check the usages in `InternalResourceGroup`.

File: presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/FairSharingQueue.java [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+        for (Node<E> candidate : candidates) {
+            totalShare += candidate.getShare();
+            totalUtilization += candidate.getUtilization();
+        }
+
+        List<Node<E>> winners = new ArrayList<>();
+        double winnerFactor = 1;
+
+        for (Node<E> candidate : candidates) {
+            double share = 1.0 * candidate.getUtilization() / totalUtilization;
```
Comment (nezihyigitbasi) [inline presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/FairSharingQueue.java]: `share` -> `actualShare` or `actualUsage` or `actualSth`. Then you compare it against `expectedSth` and you calculate the `deficit`/`delta`/`diff`.

File: presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/FairSharingQueue.java [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+            double expectedShare = 1.0 * candidate.getShare() / totalShare;
+            double candidateFactor = share / expectedShare;
+
+            if (candidateFactor <= winnerFactor) {
+                if (candidateFactor < winnerFactor) {
+                    winnerFactor = candidateFactor;
+                    winners.clear();
+                }
+                winners.add(candidate);
```
Comment (nezihyigitbasi) [inline presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/FairSharingQueue.java]: adding a comment here that explains the starvation issue this code path solves will be useful.

File: presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/Queue.java [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.execution.resourceGroups;
+
+interface Queue<E>
```
Comment (nezihyigitbasi) [inline presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/Queue.java]: this looks similar to `java.util.Queue`, please take a look to see whether we can use the sdk `Queue`.

File: presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/FairSharingQueue.java [hunk 10]
```diff
@@ -0,0 +1,196 @@
+/*
```
Comment (nezihyigitbasi) [inline presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/FairSharingQueue.java]: Please add unit tests for this new queue implementation. There is a bunch of tests for the other queue impls.

File: presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/FairSharingQueue.java [hunk 11]
```diff
@@ -0,0 +1,196 @@
...
+        if (winners.size() == 1) {
+            return winners.get(0).getValue();
+        }
+
+        long totalShares = winners.stream()
+                .mapToLong(Node::getShare)
+                .sum();
+
+        long winningTicket = ThreadLocalRandom.current().nextLong(0, totalShares);
+        long count = 0;
```
Comment (nezihyigitbasi) [inline presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/FairSharingQueue.java]: `cumulativeShare` instead of `count`?

File: presto-spi/src/main/java/com/facebook/presto/spi/resourceGroups/SchedulingPolicy.java (hunk 2 above)
Comment (nezihyigitbasi) [reply presto-spi/src/main/java/com/facebook/presto/spi/resourceGroups/SchedulingPolicy.java]: This new heuristic looks like a fair scheduler that takes into account the configured `schedulingWeight`, so how about naming it `WEIGHTED_FAIR`?

File: presto-main/src/test/java/com/facebook/presto/execution/resourceGroups/TestResourceGroups.java [hunk 12]
```diff
@@ -390,6 +385,153 @@ public void testWeightedScheduling()
...
+        BinomialDistribution binomial = new BinomialDistribution(4000, 1.0 / 4.0);
+        int lowerBound = binomial.inverseCumulativeProbability(0.000001);
+        int upperBound = binomial.inverseCumulativeProbability(0.999999);
+
+        assertBetweenInclusive(group1Ran, lowerBound, upperBound);
+        assertBetweenInclusive(group2Ran, lowerBound, upperBound);
+        assertBetweenInclusive(group3Ran, 2 * lowerBound, 2 * upperBound);
+    }
+
+    @Test(timeOut = 10_000_000) //TODO: fix
```
Comment (nezihyigitbasi) [inline presto-main/src/test/java/com/facebook/presto/execution/resourceGroups/TestResourceGroups.java]: what's this TODO for? is it related to extremely high timeout?

File: presto-main/src/test/java/com/facebook/presto/execution/resourceGroups/TestResourceGroups.java [hunk 13]
```diff
@@ -390,6 +385,153 @@ public void testWeightedScheduling()
...
+        int group1Ran = 0;
+        for (int i = 0; i < 2000; i++) {
+            group1Ran += completeGroupQueries(group1Queries);
+            completeGroupQueries(group2Queries);
+            root.processQueuedQueries();
+            group1Queries = fillGroupTo(group1, group1Queries, 4);
+            group2Queries = fillGroupTo(group2, group2Queries, 4);
+        }
+
+        // group 3 should run approximately 2x the number of queries of 1 and 2
```
Comment (nezihyigitbasi) [inline presto-main/src/test/java/com/facebook/presto/execution/resourceGroups/TestResourceGroups.java]: there is no group 3 here ? What's this test really testing?

File: presto-spi/src/main/java/com/facebook/presto/spi/resourceGroups/SchedulingPolicy.java (hunk 2 above)
Comment (raghavsethi) [reply presto-spi/src/main/java/com/facebook/presto/spi/resourceGroups/SchedulingPolicy.java]: That makes it sound related to 'FAIR', which it isn't at all..

File: presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/Queue.java (hunk 9 above)
Comment (raghavsethi) [reply presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/Queue.java]: The java.util.Queue is much too heavyweight. We would have to throw UnsupportedOperationException for a half dozen methods.

File: presto-main/src/test/java/com/facebook/presto/execution/resourceGroups/TestResourceGroups.java (hunk 12 above)
Comment (raghavsethi) [reply presto-main/src/test/java/com/facebook/presto/execution/resourceGroups/TestResourceGroups.java]: Unintended, reverted.

File: presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/FairSharingQueue.java (hunk 6 above)
Comment (raghavsethi) [reply presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/FairSharingQueue.java]: Yes, this is part of the API.

File: presto-spi/src/main/java/com/facebook/presto/spi/resourceGroups/SchedulingPolicy.java (hunk 2 above)
Comment (nezihyigitbasi) [reply presto-spi/src/main/java/com/facebook/presto/spi/resourceGroups/SchedulingPolicy.java]: I still think `WEIGHTED_FAIR` is a better fit, but I leave the naming part to you guys to decide.
